url24d793285d_auzll_gengweibo_tgz-pJ8-org_json_TestJ8/src/net/oauth/http/HttpMessage.java

```diff
@@ -0,0 +1,221 @@
+import java.util.regex.Pattern;
+import net.oauth.client.ExcerptInputStream;
+import net.oauth.OAuth;
+import net.oauth.OAuthMessage;
+import net.oauth.OAuthProblemException;
+import net.oauth.ParameterStyle;
+
+/**
+ * An HTTP request or response.
+ * 
+ * @author John Kristian
+ */
+public class HttpMessage
+{
+
+    public HttpMessage()
+    {
+        this(null, null);
+    }
+
+    public HttpMessage(String method, URL url)
+    {
+        this(method, url, null);
+    }
+
+    public HttpMessage(String method, URL url, InputStream body)
+    {
+        this.method = method;
+        this.url = url;
+        this.body = body;
+    }
+
+    public String method;
+    public URL url;
+    public final List<Map.Entry<String, String>> headers = new ArrayList<Map.Entry<String, String>>();
+    protected InputStream body = null;
+
+    /**
+     * Get the value of the last header of the given name. The name is
+     * case-insensitive.
+     */
+    public final String getHeader(String name)
+    {
+        String value = null;
+        for (Map.Entry<String, String> header : headers) {
+            if (equalsIgnoreCase(name, header.getKey())) {
+                value = header.getValue();
+            }
+        }
+        return value;
+    }
+
+    /**
+     * Remove all headers of the given name. The name is case insensitive.
+     * 
+     * @return the value of the last header with that name, or null to indicate
+     *         there was no such header
+     */
+    public String removeHeaders(String name)
+    {
+        String value = null;
+        for (Iterator<Map.Entry<String, String>> i = headers.iterator(); i.hasNext();) {
+            Map.Entry<String, String> header = i.next();
+            if (equalsIgnoreCase(name, header.getKey())) {
+                value = header.getValue();
+                i.remove();
+            }
+        }
+        return value;
+    }
+
+    public final String getContentCharset()
+    {
+        return getCharset(getHeader(CONTENT_TYPE));
+    }
+
+    public final InputStream getBody() throws IOException
+    {
+        if (body == null) {
+            InputStream raw = openBody();
+            if (raw != null) {
+                body = new ExcerptInputStream(raw);
+            }
+        }
+        return body;
+    }
+
+    protected InputStream openBody() throws IOException
+    {
+        return null;
+    }
+
+    /** Put a description of this message and its origins into the given Map. */
+    public void dump(Map<String, Object> into) throws IOException
+    {
+    }
+
+    /**
+     * Construct an HTTP request from this OAuth message.
+     * 
+     * @param style
+     *            where to put the OAuth parameters, within the HTTP request
+     */
+    public static HttpMessage newRequest(OAuthMessage from, ParameterStyle style) throws IOException {
+        final boolean isPost = OAuthMessage.POST.equalsIgnoreCase(from.method);
+        InputStream body = from.getBodyAsStream();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *body*. Currently, there are execution paths that do not contain closure statements, for example, when OAuth.formEncode throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).